### PR TITLE
sql/pgwire: send down parameter status updates

### LINKED ISSUE #40854: sql: set ApplicationName from JDBC
The PGJDBC tests connect to the database with a connection string that looks like this: `jdbc:postgresql://localhost:26257/defaultdb?ApplicationName=Driver Tests`

Some of the tests then expect the `application_name` parameter to be `Driver Tests`. For example, see ParameterStatusTest.java in the pgjdbc repo.

It also seems like these don't work properly:

```
    con.setClientInfo("ApplicationName", "my app");
    assertEquals("my app", con.getClientInfo("ApplicationName"));

    s.execute("set application_name='" + appName + "'");
    assertEquals("application_name was set to " + appName + ", and it should be visible via "
        + "con.getClientInfo", appName, con.getClientInfo("ApplicationName"));

    Properties props = new Properties();
    props.put("ApplicationName", "my app");
    con.setClientInfo(props);
    assertEquals("my app", con.getClientInfo("ApplicationName"));
```

### PR DESCRIPTION
**Pull request by otan**
Resolves #40854.

Server parameters such as `application_name` and `timezone` require updates
over pgwire when they are changed. This was previously not done.

In this PR, we introduce a listener-esque object on `sessionDataMutator`
that pgwire will add itself onto to send updates on these changes.

Coincidentally, this means a few logic tests changes because lib/pq will
encode this into the string.

Release note (sql change): This PR introduces a new pgwire update that
sends ParameterStatus messages when certain server parameters are
changed for the given session over pgwire.

### CODE REVIEW

File: pkg/sql/pgwire/conn.go [hunk 1]
```diff
@@ -610,10 +610,18 @@ func (c *conn) sendInitialConnData(
...
 	// For details see: https://www.postgresql.org/docs/10/static/libpq-status.html
 	for _, param := range statusReportParams {
+		param := param
 		value := connHandler.GetStatusParam(ctx, param)
 		if err := c.sendStatusParam(param, value); err != nil {
 			return sql.ConnectionHandler{}, err
 		}
+		// pgwire also expects updates when these parameters change.
+		connHandler.RegisterOnSessionDataChange(param, func(val interface{}) {
+			convertVal := val.(string)
```
Comment (mjibson) [inline pkg/sql/pgwire/conn.go]: Can val be a `string` instead of `interface{}` to avoid the forced cast? Seems unsafe otherwise.

**Review by madelynnblue — COMMENTED** [review]
Does this make that JDBC test pass?

**Review by otan — COMMENTED** [review]
Yep, the tests pass with this change.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42376#-:-LtR_V9p5xsvvLIwcXqe:b43aq14)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson, @otan, and @rafiss)

___
*[pkg/sql/exec_util.go, line 1743 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42376#-LtRJ3WgJzylt1pbu3tA-r1-1743:-LtR_c-l16pA_QVNv0_y:bd3j9ys) ([raw file](https://github.com/cockroachdb/cockroach/blob/98b1e7b845763211dc69a9ea6ec4d7d7825cda6b/pkg/sql/exec_util.go#L1743)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Why does this need to be a slice? Can there ever by more than one per key?
</blockquote></details>

Yes, I think the `application_name` one can be used more than once - once for stats and once for pgwire (the change for this is somewhere else in this PR).
___
*[pkg/sql/exec_util.go, line 1752 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42376#-LtRJ3WgJzylt1pbu3tA-r1-1752:-LtR_msA57KRe8dVpZfn:bb71npn) ([raw file](https://github.com/cockroachdb/cockroach/blob/98b1e7b845763211dc69a9ea6ec4d7d7825cda6b/pkg/sql/exec_util.go#L1752)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

There's no need for this block.
</blockquote></details>

Ooh.
___
*[pkg/sql/pgwire/conn.go, line 620 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42376#-LtRJ3WgJzylt1pbu3tB-r1-620:-LtR_p_zERF61AtwKqaE:b-mjd870) ([raw file](https://github.com/cockroachdb/cockroach/blob/98b1e7b845763211dc69a9ea6ec4d7d7825cda6b/pkg/sql/pgwire/conn.go#L620)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Can val be a `string` instead of `interface{}` to avoid the forced cast? Seems unsafe otherwise.
</blockquote></details>

Yes, I saw some `Set` params that were not `string`s, but they don't seem to need listeners (yet) so okay with doing this.
___
*[pkg/sql/pgwire/testdata/pgtest/connection_params, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42376#-LtRJ3WgJzylt1pbu3tC-r1-22:-LtRb10i-38kYNf3UCgW:b-7gclag) ([raw file](https://github.com/cockroachdb/cockroach/blob/98b1e7b845763211dc69a9ea6ec4d7d7825cda6b/pkg/sql/pgwire/testdata/pgtest/connection_params#L22)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Do we support word time zones (`MDT`)? If so I'd like to see another test with that.
</blockquote></details>

Added a test.

**Review by otan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42376)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson and @rafiss)

___
*[pkg/sql/pgwire/testdata/pgtest/connection_params, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42376#-LtRJ3WgJzylt1pbu3tC-r1-1:-LtRcO314rUUB5pr7t4n:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/98b1e7b845763211dc69a9ea6ec4d7d7825cda6b/pkg/sql/pgwire/testdata/pgtest/connection_params#L1)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

If this file differs from what postgres does it should document that with a comment.
</blockquote></details>

Done.

File: pkg/sql/pgwire/testdata/pgtest/connection_params [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ReadyForQuery
+----
+{"Type":"ParameterStatus","Name":"application_name","Value":"pgtest"}
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"CommandComplete","CommandTag":"SET"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+# Change the time zone.
+# NOTE: the order is different than postgres, because we execute the statement
```
Comment (mjibson) [inline pkg/sql/pgwire/testdata/pgtest/connection_params]: once at the top is enough

File: pkg/sql/pgwire/testdata/pgtest/connection_params [hunk 3]
```diff
@@ -0,0 +1,55 @@
+# Change the application name.
+# NOTE: the order is different than postgres, because we execute the statement
```
Comment (mjibson) [inline pkg/sql/pgwire/testdata/pgtest/connection_params]: Are you sure this is the reason? The commit message is more unsure. (I'm also unsure, I just don't want to have an incorrect statement here.)

**Review by madelynnblue — COMMENTED** [review]
Can you flesh out the comment in the commit message about exactly why the logic test results changed? Is it because lib/pq supports reading the timezone back from the connection and sets the location data of received timestamps from it?

**Review by otan — COMMENTED** [review]
> Can you flesh out the comment in the commit message about exactly why the logic test results changed? Is it because lib/pq supports reading the timezone back from the connection and sets the location data of received timestamps from it?
yep!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42376#-:-LtV2Hgx7yAVdbyUE4DC:b6dgdfj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson and @rafiss)

___
*[pkg/sql/pgwire/testdata/pgtest/connection_params, line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42376#-LtRJ3WgJzylt1pbu3tC-r2-2:-LtV2LxnEYz6U7KYKiYk:b-75yavf) ([raw file](https://github.com/cockroachdb/cockroach/blob/db63b33309dcdf02c2d7825032e455a0702db562/pkg/sql/pgwire/testdata/pgtest/connection_params#L2)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Are you sure this is the reason? The commit message is more unsure. (I'm also unsure, I just don't want to have an incorrect statement here.)
</blockquote></details>

I dug even further into this and realised I wasn't buffering. I got mislead earlier.
___
*[pkg/sql/pgwire/testdata/pgtest/connection_params, line 22 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42376#-LtRJ3WgJzylt1pbu3tC-r2-22:-LtV2QV_7_s6PcmqC-z6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db63b33309dcdf02c2d7825032e455a0702db562/pkg/sql/pgwire/testdata/pgtest/connection_params#L22)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

once at the top is enough
</blockquote></details>

Done.

File: pkg/sql/vars.go [hunk 4]
```diff
@@ -702,16 +702,7 @@ var varGen = map[string]sessionVar{
 	// See https://www.postgresql.org/docs/10/static/runtime-config-client.html#GUC-TIMEZONE
 	`timezone`: {
 		Get: func(evalCtx *extendedEvalContext) string {
-			// If the time zone is a "fixed offset" one, initialized from an offset
```
Comment (otan) [inline pkg/sql/vars.go]: (this looks already broken, so i will fix this in a separate PR - if you have a better name idea though lmk!)

**Comment by rafiss** [comment]
Nice stuff! Could you take a look at `pgjdbc_blacklist.go`, then find the tests that are failing due to issue #40854, and update the blacklist for v20.1 if the tests pass now?

File: pkg/sql/vars.go (hunk 4 above)
Comment (mjibson) [reply pkg/sql/vars.go]: That's fine. But why change it in this PR?

Comment (otan) [reply pkg/sql/vars.go]: I needed this function here: https://github.com/cockroachdb/cockroach/pull/42376/files#diff-4fd6eaee39da7af19c97fbf9311f2c1fR1832

**Review by madelynnblue — APPROVED** [review]

**Comment by otan** [comment]
thanks!

bors r+

**Review by andreimatei — COMMENTED** [review]
@otan did you attempt to implement this through the `sql.CommandResult` interface, instead of creating a new communication mechanism between sql and pgwire? I'm thinking the `sql.CommandResult`could carry these updates, which would make the control flow much easier to reason about.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42376#-:-M-Vxt8uA14Kgz9mD-7r:bjm6zza)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson and @rafiss)

**Comment by otan** [comment]
@andreimatei that works too, see https://github.com/cockroachdb/cockroach/pull/44883. couldn't get this to work for `CommandResult`, but did something similar. This will be a nice extension for the work needed for notices.

**Pull request merged**